### PR TITLE
Specify the operand data type constraints of operations

### LINKED ISSUE #283: Specify the operand data type constraints of operation
The current spec doesn't specify the operand type constraints of an operation. However, some operations, e.g., [`softmax`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-softmax) should only support float32 operand type according to the survey of frameworks and native ML APIs in the following table. The lack of the operand type constraints specification would lead to implementation issue, such as [Chromium CL 3856752](https://chromium-review.googlesource.com/c/chromium/src/+/3856752/comments/267c926c_b14af87f).

| Framework / API | supported data types |
|---|---|
| TensorFlow [tf.nn.softmax](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) | half, float32, float64 |
| ONNX [Softmax](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Softmax) | tensor(float16), tensor(float), tensor(double), tensor(bfloat16) |
| NNAPI [ANEURALNETWORKS_SOFTMAX](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a2bfbb83a537701e2843a3d5004250c2c) | FLOAT32, FLOAT16 |
| DirectML [DML_ACTIVATION_SOFTMAX_OPERATOR_DESC](https://docs.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_activation_softmax_operator_desc) | FLOAT32, FLOAT16 |
| BNNS [ActivationFunction](https://developer.apple.com/documentation/accelerate/bnns/activationlayer/3649741-init) | float |

Thanks @wacky6 for pointing this out.

### PR DESCRIPTION
**Pull request by inexorabletash**
Introduce constraints for input operands, either directly (e.g. input's dataType can only be "float32" or "float16") or indirectly (e.g. weight's dataType must be the same as input's).

Fixes #283

***
<a href="https://pr-preview.s3.amazonaws.com/inexorabletash/webnn/pull/646.html" title="Last updated on Apr 26, 2024, 6:03 AM UTC (ad239bc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/646/d57e4ef...inexorabletash:ad239bc.html" title="Last updated on Apr 26, 2024, 6:03 AM UTC (ad239bc)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -3240,12 +3257,17 @@ partial interface MLGraphBuilder {
...
     1. If |options|.{{MLGruCellOptions/activations}} [=map/exists=], and [=MLGraphBuilder/validating activation=] with [=this=] and any [=list/item=] in it returns false, then [=exception/throw=] a {{TypeError}}.
+    1. If |input|'s [=MLOperand/dataType=] is not {{MLOperandDataType/"float32"}} or {{MLOperandDataType/"float16"}}, then [=exception/throw=] a {{TypeError}}.
     1. If the [=MLOperand/rank=] of any of |input|, |weight|, |recurrentWeight| or |hiddenState| is not 2, then [=exception/throw=] a {{TypeError}}.
+    1. If the [=MLOperand/dataType=] of either |weight| or |recurrentWeight| is not equal to |input|'s [=MLOperand/dataType=], then [=exception/throw=] a {{TypeError}}.
     1. If |weight|'s [=MLOperand/shape=][0] is not equal to 3 * |hiddenSize|, then [=exception/throw=] a {{TypeError}}.
     1. If |recurrentWeight|'s [=MLOperand/shape=][0] is not equal to 3 * |hiddenSize|, then [=exception/throw=] a {{TypeError}}.
+    1. If |hiddenState|'s [=MLOperand/dataType=] is not equal to |input|'s [=MLOperand/dataType=], then [=exception/throw=] a {{TypeError}}.
```
Comment (huningxin) [inline index.bs:3265]: Could `hiddenState` be merged into the validation for `weight` and `recurrentWeight`? Line 3262

**Review by huningxin — APPROVED** [review]
LGTM with a nit, thanks much!

File: index.bs (hunk 1 above)
Comment (inexorabletash) [reply index.bs]: Good catch! Done in 1d91d80bf53f3a43b8cceb7342b324a4abdcdb9a

**Review by huningxin — APPROVED** [review]
LGTM, thanks!

File: index.bs [hunk 2]
```diff
@@ -2489,91 +2497,91 @@ partial interface MLGraphBuilder {
...
-        1. Let |output| be the result of running the [=MLGraphBuilder/element-wise-unary-op | create element-wise unary operation=] given "floor" and |input|.
+        1. Let |output| be the result of running the [=MLGraphBuilder/element-wise-unary-op | create element-wise unary operation=] given "floor", |input|, and « {{MLOperandDataType/"float32"}}, {{MLOperandDataType/"float16"}} ».
             1. If that [=exception/throws=] an error, then re-[=exception/throw=] the error.
         1. Return |output|.
     </div>
 
     <div algorithm>
     The <dfn method for=MLGraphBuilder>identity(|input|)</dfn> method steps are:
-        1. Let |output| be the result of running the [=MLGraphBuilder/element-wise-unary-op | create element-wise unary operation=] given "identity" and |input|.
+        1. Let |output| be the result of running the [=MLGraphBuilder/element-wise-unary-op | create element-wise unary operation=] given "identity", |input|, and « {{MLOperandDataType/"float32"}}, {{MLOperandDataType/"float16"}} ».
```
Comment (fdwr) [inline index.bs:2542]: Identity should accept all types - it does no math and just copies data forward.

Comment (inexorabletash) [reply index.bs]: Good catch! Fixed in 16e424cff421762aed9d40ec826b0f1d7cab72b5

**Review by huningxin — APPROVED** [review]
LGTM, thanks for the fix!

**Comment by inexorabletash** [comment]
Conflict resolved. @fdwr, good to merge?

**Comment by inexorabletash** [comment]
I'd missed reduceMean() somehow. Addressed in bd9b0cfa85df6798c7b83c849bf8107501d8ee4a

**Review by huningxin — APPROVED** [review]
LGTM

File: index.bs [hunk 3]
```diff
@@ -2948,6 +2958,8 @@ partial interface MLGraphBuilder {
     The <dfn method for=MLGraphBuilder>gemm(|a|, |b|, |options|)</dfn> method steps are:
   </summary>
     1. If [=MLGraphBuilder/validating operand=] with [=this=] and any of |a| and |b| returns false, then [=exception/throw=] a {{TypeError}}.
+    1. If |a|'s [=MLOperand/dataType=] is not {{MLOperandDataType/"float32"}} or {{MLOperandDataType/"float16"}}, then [=exception/throw=] a {{TypeError}}.
+    1. If |b|'s [=MLOperand/dataType=] is not equal to |a|'s [=MLOperand/dataType=], then [=exception/throw=] a {{TypeError}}.
     1. If |a|'s [=MLOperand/rank=] is not 2 or |b|'s [=MLOperand/rank=] is not 2, then [=exception/throw=] a {{TypeError}}.
     1. Let |shapeA| be a [=list/clone=] of |a|'s [=MLOperand/shape=].
     1. Let |shapeB| be a [=list/clone=] of |b|'s [=MLOperand/shape=].
```
[the remark this answers is not in the document]
Comment (fdwr) [reply index.bs:2965]: I just did it. Please review my commit.

**Review by fdwr — APPROVED** [review]
Thanks JB. Tedious work there! I update GEMM C's data type check (please review). The rest LGTM, and my ideas can wait.

File: index.bs [hunk 4]
```diff
@@ -4923,21 +4971,21 @@ partial interface MLGraphBuilder {
 
     <div algorithm>
     The <dfn method for=MLGraphBuilder>reduceProduct(|input|, |options|)</dfn> method steps are:
-        1. Let |output| be the result of running the [=MLGraphBuilder/reduce-op | create reduce operation=] given "reduceProduct", |input| and |options|.
+        1. Let |output| be the result of running the [=MLGraphBuilder/reduce-op | create reduce operation=] given "reduceProduct", |input|, |options|, and « {{MLOperandDataType/"float32"}}, {{MLOperandDataType/"float16"}}, {{MLOperandDataType/"int32"}}, {{MLOperandDataType/"uint32"}} ».
```
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:4973]: Yeah, I raised this over in https://github.com/webmachinelearning/webnn/pull/643#discussion_r1569231289  - i.e. using the list style everywhere. I'm a fan and happy to change things to this form everywhere:

* If input’s dataType is not in « "float32", "float16" », then ...
* If input’s dataType is not in « "float32", "float16", "int32", "int8" », then ...

... with the question of which formulation to use if there's a single type:

* If input’s dataType is not "uint8", then ...
* If input’s dataType is not « "uint8" », then ...

Thoughts?

File: index.bs [hunk 5]
```diff
@@ -4385,6 +4428,8 @@ partial interface MLGraphBuilder {
     The <dfn method for=MLGraphBuilder>matmul(|a|, |b|)</dfn> method steps are:
   </summary>
     1. If [=MLGraphBuilder/validating operand=] with [=this=] and any of |a| and |b| returns false, then [=exception/throw=] a {{TypeError}}.
+    1. If |a|'s [=MLOperand/dataType=] is not {{MLOperandDataType/"float32"}} or {{MLOperandDataType/"float16"}}, then [=exception/throw=] a {{TypeError}}.
+    1. If |b|'s [=MLOperand/dataType=] is not equal to |a|'s [=MLOperand/dataType=], then [=exception/throw=] a {{TypeError}}.
```
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:4431]: Let me try a parallel CR to this one that tries this. I have a love/hate relationship with the Arguments/Returns preamble; since it's not really rigorous I think of it as non-normative. Listing types out-of-band will need some formal structure, and e.g. "same as a" needs to be precisely defined (i.e. the actual data type needs to match)

File: index.bs (hunk 3 above)
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:2965]: LGTM

File: index.bs (hunk 4 above)
[the remark this answers is not in the document]
Comment (fdwr) [reply index.bs:4973]: I guess a type list of 1 type is still a type list. So for consistency imo, `If input’s dataType is not « "uint8" », then ...`.

File: index.bs (hunk 5 above)
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:4431]: I've gotten some feedback from @a-sully and @reillyeon about the approaches, and what this could enable (e.g. merge w/ Arguments, make it define shape as well, ...). 

I might suggest that we merge this PR, either as-is or with the [everything is a list](https://github.com/webmachinelearning/webnn/pull/646#discussion_r1581284024) change suggested above, and then iterate on table approach. Having this PR together has already identified some [implementation gaps](https://g-issues.chromium.org/issues/328567884#comment2) so start with this then iterate?

[the remark this answers is not in the document]
Comment (fdwr) [reply index.bs:4431]: > I might suggest that we merge this PR, either as-is 

Yep yep. I'll do that now.

> make it define shape as well

That would be interesting. Some other specs I've read were hand-wavy about how the output shape calculations occur, leaving a lot of room for ~~bugs~~ interpretation in implementations. So, seeing that clearly conveyed, without needing to refer to some separate Javascript codebase, would be nice.
